nav2_costmap_2d/plugins/denoise_layer.cpp

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) 2021 Andrey Ryzhikov
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_costmap_2d/denoise_layer.hpp"
+
+#include <string>
+#include <vector>
+#include <algorithm>
+
+#include "rclcpp/rclcpp.hpp"
+
+namespace nav2_costmap_2d
+{
+bool isBackground(uint8_t pixel)
+{
+  return pixel != LETHAL_OBSTACLE && pixel != NO_INFORMATION;
+}
+
+void
+DenoiseLayer::onInitialize()
+{
+  // Enable/disable plugin
+  declareParameter("enabled", rclcpp::ParameterValue(true));
+  // Smaller groups should be filtered
+  declareParameter("minimal_group_size", rclcpp::ParameterValue(2));
+  // Pixels connectivity type
+  declareParameter("group_connectivity_type", rclcpp::ParameterValue(8));
+  // Max additional memory amount
+  declareParameter("memory_usage_mb", rclcpp::ParameterValue(0));
+
+  const auto node = node_.lock();
+
+  if (!node) {
+    throw std::runtime_error("DenoiseLayer::onInitialize: Failed to lock node");
+  }
+  node->get_parameter(name_ + "." + "enabled", enabled_);
+
+  auto getInt = [&](const std::string & parameter_name) {
+      int param{};
+      node->get_parameter(name_ + "." + parameter_name, param);
+      return param;
+    };
+
+  const int minimal_group_size_param = getInt("minimal_group_size");
+
+  if (minimal_group_size_param <= 1) {
+    RCLCPP_WARN(
+      logger_,
+      "DenoiseLayer::onInitialize(): param minimal_group_size: %i."
+      " A value of 1 or less means that all map cells will be left as they are.",
+      minimal_group_size_param);
+    this->minimal_group_size_ = 1;
```

Comment:
Remove the `this->`'s, the trailing underscore should make it clear its a class member